### PR TITLE
Added model_classes

### LINKED ISSUE #2056: [FEAT] Make class mechanism in NLP match the one in tabular & vision
**Is your feature request related to a problem? Please describe.**
Following https://github.com/deepchecks/deepchecks/pull/2050, scorers receive the list of classes in the task from the context. 
- In tabular, the context aggregates the class from: the .classes_ in the model, the labels in the data, the predicted labels OR the user supplied list of classes (passed to the Dataset constructor).
- In vision, the context gets the classes from proba shape (always 0..n-1)

So now the NLP branch is out of sync with this change and must pass it's own classes to the DeepcheckScorer constructed in nlp scorers.py. Need to implement a mechanism in the context to get the list of all classes in the task, similarly to what is done in tabular and vision.

**Describe the solution you'd like**
On the face of it it should be implemented like it is in tabular - look at all present classes in labels and prediction, OR just receive classes from the user through TextData constructor. The complication is that in NLP we also implemented multilabel, whose label format (binary matrices) requires the possibility of passing a "label dict" like the one we have in vision, giving names to the classes from 0 to n-1. This is the current use of the `classes` argument of the TextData constructor. 

So we have a complicated situation here - in the case of regular classification / token classification, labels are like in tabular (the label you see in the data is the label name), but for multilabel tasks you must find a way to let deepchecks know what are the class names.

**goal**
Suggest a solution to that will catch all these cases and enable us to pass a list of classes to the constructor of DeepcheckScorer . Refactor the TextData `classes` argument and NLP context to implement suggestion.

### PR DESCRIPTION
**Pull request by nirhutnik**
@noamzbr check my questions

closes #2056

### CODE REVIEW

File: deepchecks/nlp/base_checks.py [hunk 1]
```diff
@@ -37,6 +37,7 @@ def run(
         with_display: bool = True,
         predictions: Optional[TTextPred] = None,
         probabilities: Optional[TTextProba] = None,
+        model_classes: Optional[Union[List, List[List]]] = None,
```
Comment (noamzbr) [inline deepchecks/nlp/base_checks.py:40]: When is that a list of lists?

File: deepchecks/nlp/context.py [hunk 2]
```diff
@@ -295,6 +299,7 @@ def __init__(
             test_pred: t.Optional[TTextPred] = None,
             train_proba: t.Optional[TTextProba] = None,
             test_proba: t.Optional[TTextProba] = None,
+            model_classes: t.Optional[t.List] = None,
```
Comment (noamzbr) [inline deepchecks/nlp/context.py:302]: here it's just a list!

File: deepchecks/nlp/utils/data_inference.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            train_labels += train_dataset.label
+        if have_model:
+            train_labels += model.predict(train_dataset)
+    if test_dataset:
+        if test_dataset.has_label():
+            test_labels += test_dataset.label
+        if have_model:
+            test_labels += model.predict(test_dataset)
+
+    observed_labels = np.array(test_labels + train_labels)
```
Comment (noamzbr) [inline deepchecks/nlp/utils/data_inference.py:53]: Does this whole code work for token classification?

File: deepchecks/nlp/utils/data_inference.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    if len(observed_labels.shape) == 2:
+        observed_labels = [np.unique(observed_labels[:, i]) for i in range(observed_labels.shape[1])]
+        observed_labels = [x[~pd.isnull(x)] for x in observed_labels]
+        observed_labels = [sorted(x) for x in observed_labels]
+        # Bressler - this is more generalized for multilabel, instead of being specific for [0,1] labels. Should keep?
```
Comment (noamzbr) [inline deepchecks/nlp/utils/data_inference.py:58]: Not sure I follow. Our supported multilabel label format is a binary matrix (for which I didn't manage to follow how this code will work). What is generalized multilabel?

File: deepchecks/nlp/text_data.py [hunk 5]
```diff
@@ -292,7 +292,7 @@ def label(self) -> TTextLabel:
         return self._label
 
     @property
-    def classes(self) -> t.Optional[t.List[t.Union[str, int]]]:
+    def classes(self) -> t.Optional[t.List[t.Union[str, int]]]:  # Bressler - should keep or use context.model_classes?
```
Comment (noamzbr) [inline deepchecks/nlp/text_data.py:295]: Seems like if there is a model_classes argument in run() then there is no need for a classes argument for TextData. This means that for multilabel, model_classes should receive the names the classes should have (because again in multilabel we have a binary matrix so we don't know what the labels are)
Following that logic the logic here for token classification should go into the infer_observed_labels method, no?

File: tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py [hunk 6]
```diff
@@ -108,6 +108,41 @@ def test_run_with_scorer_multilabel_class_names(text_multilabel_classification_d
...
+    # Arrange
+    text_multilabel_classification_dataset_mock_copy = text_multilabel_classification_dataset_mock.copy()
+    text_multilabel_classification_dataset_mock_copy._classes = ['a', 'b', 'c']
+    text_multilabel_classification_dataset_mock_copy._label = [[0, 0, 1], [0, 1, 0], [0, 1, 0]]
+    check = SingleDatasetPerformance(scorers=['f1_per_class'])
+
+    # Act
+    result = check.run(text_multilabel_classification_dataset_mock_copy,
+                       predictions=[[0, 0, 1], [1, 0, 1], [0, 1, 0]],
+                       model_classes=[[0, 1], [0, 1], [0, 1]]
```
Comment (noamzbr) [inline tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py:121]: As said in the comments above, IMO model_classes should be the one conveying the info that the class names are `['a', 'b', 'c']`

File: deepchecks/nlp/utils/data_inference.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        if have_model:
+            train_labels += model.predict(train_dataset)
+    if test_dataset:
+        if test_dataset.has_label():
+            test_labels += test_dataset.label
+        if have_model:
+            test_labels += model.predict(test_dataset)
+
+    observed_classes = np.array(test_labels + train_labels)
+    if len(observed_classes.shape) == 2:
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/utils/data_inference.py:64]: done 👍

File: deepchecks/nlp/context.py [hunk 8]
```diff
@@ -317,11 +324,25 @@ def __init__(
...
                                          'dataset, initialize it as train_dataset')
+        if model_classes and len(model_classes) == 0:
+            raise DeepchecksValueError('Received empty model_classes')
+        if model_classes and sorted(model_classes) != model_classes:
+            supported_models_link = doclink(
+                'nlp-supported-predictions-format',
+                template='For more information please refer to the Supported Tasks guide {link}')
+            raise DeepchecksValueError(f'Received unsorted model_classes. {supported_models_link}')
+
+        self._task_type = train_dataset.task_type if train_dataset else test_dataset.task_type if test_dataset else None
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/context.py:335]: Well... Then fuck? That shouldn't happen anyway.

File: deepchecks/nlp/context.py [hunk 9]
```diff
@@ -266,21 +270,23 @@ class Context(BaseContext):
 
     Parameters
     ----------
-    train_dataset: Union[TextData, None] , default: None
+    train_dataset : Union[TextData, None], default: None
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/context.py:273]: Come onnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn

File: deepchecks/nlp/context.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/context.py:335]: Can add a simple Deepchecks error for that

File: deepchecks/nlp/context.py [hunk 10]
```diff
@@ -141,7 +147,7 @@ def __init__(self,
             self._proba_indices = \
                 {name: set(data_proba.keys()) for name, data_proba in self.probas.items()}
 
-    def _predict(self, data: TextData) -> TTextPred:
+    def _predict(self, data: TextData) -> TTextPred:  # TODO: Needs to receive list of strings, not TextData
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/context.py:150]: No need for issue, small fix will be done in some next task. Won't be left there

File: deepchecks/nlp/context.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/context.py:335]: But I don't want to

[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/context.py:335]: Done

File: deepchecks/nlp/context.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/context.py:273]: Done

**Review by noamzbr — APPROVED** [review]
